### PR TITLE
docs: model: tensorflow_hub: Add example for text classifier.

### LINKED ISSUE #435: docs: model: tensorflow hub: Include examples of how to use from Python API
The model plugins page has command line examples but needs examples from Python as well.

Copy the format of the `examples/quickstart_filenames.py` file and add a `.. code-block:: python` section for each model, keeping the Python example consistent with the command line example for each model. Let's add the python example after the CLI example for each.

For the tensorflow hub model, the docstring of the `Model` needs to be updated.

**Edit:**

Here's a good example from when tensorflow was updated: https://github.com/intel/dffml/commit/e7aaeb1e1e2f1b2ce74d1fe6ecc6422bf6df178c

### LINKED ISSUE #537: Tensorflow_hub / scratch: ubuntu: python api does not run successfully as scikit does
## 'module' object is not callable

This error shows when you run scratch or tensorflow hub models in python. But works perfectly for scikit

## Steps To Reproduce

```python
from dffml import CSVSource, Features, DefFeature
from dffml.noasync import train, accuracy, predict
from dffml_model_tensorflow_hub import text_classifier


model = text_classifier(
features = Features(
DefFeature("Sentence" , str , 1),
),
    predict = DefFeature('Sentiment', int ,1)

)
train(model , {'I hate you!!' , 2} , {'Okay , so what do we do now?' , 0} ,
     {'I find this dish incredibly tasty' , 1} , {'This movie is horrible' , 2} ,
     {'You look beautiful' , 1} , {'I think your project could use more collabaration' , 0}
     ,{'I cannot see any effort put here' ,2} , {'This is exactly what I wanted',  1})

print(accuracy(model,
              {'This is horrific', 2} , {'Its okay never mind' , 0} , {'This time could not have been better' , 1}))

for i, features, prediction in predict(
    model,
    {'This is horrific'} , {'Its okay never mind'} , {'This time could not have been better'}
):
    features["Sentiment"] = prediction["Sentiment"]["value"]
    print(features)
```

## Packages in use and versions

- dffml==0.3.5

### PR DESCRIPTION
**Pull request by mHash1m**
fix #435 
fix #537 
close #542

### CODE REVIEW

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+from dffml_model_tensorflow_hub.text_classifier import TextClassificationModel
+
+model = TextClassificationModel(
+    features=Features(DefFeature("Sentence", str, 1)),
+    predict=DefFeature("Sentiment", int, 1),
+    classifications=[0, 1, 2],
+    clstype=int,
+)
+# train the model
+train(model, "train.csv")
```
Comment (mHash1m) [inline model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:12]: @pdxjohnny I've followed the format used in other examples. However, I'm having trouble running this and the other examples, that you mentioned for reference, locally. It seems that it doesn't find the "train.csv" file. While running the test, it says "no records to train on". So I suppose it is the same problem that I face locally that the tests are facing.

Comment (mHash1m) [reply model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:12]: But this would also mean that the other examples for scratch etc. are also not working which seemed unlikely so I wanted to know if there is something else bugging it, before going about it.

File: model/tensorflow_hub/examples/tfhub_text_classifier/accuracy.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+dffml accuracy \\
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/accuracy.sh:1]: All those these double backslashes should be single backslashes

```suggestion
dffml accuracy \
```

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py (hunk 1 above)
Comment (pdxjohnny) [reply model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:12]: There's a missmatch on the capitalization of the headers in the csv files vs. the features defined

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from dffml import CSVSource, Features, DefFeature
+from dffml.noasync import train, accuracy, predict
+from dffml_model_tensorflow_hub.text_classifier import TextClassificationModel
+
+model = TextClassificationModel(
+    features=Features(DefFeature("Sentence", str, 1)),
+    predict=DefFeature("Sentiment", int, 1),
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:7]:
```suggestion
    predict=DefFeature("sentiment", int, 1),
```

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from dffml import CSVSource, Features, DefFeature
+from dffml.noasync import train, accuracy, predict
+from dffml_model_tensorflow_hub.text_classifier import TextClassificationModel
+
+model = TextClassificationModel(
+    features=Features(DefFeature("Sentence", str, 1)),
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:6]:
```suggestion
    features=Features(DefFeature("sentence", str, 1)),
```

File: model/tensorflow_hub/dffml_model_tensorflow_hub/text_classifier.py [hunk 5]
```diff
@@ -326,57 +326,33 @@ class TextClassificationModel(Model):
...
+
+    .. literalinclude:: /../model/tensorflow_hub/examples/tfhub_text_classifier/test_data.sh
+
+    Train the model
+
+    .. literalinclude:: /../model/tensorflow_hub/examples/tfhub_text_classifier/train.sh
+
+    Assess the accuracy
+
+    .. literalinclude:: /../model/tensorflow_hub/examples/tfhub_text_classifier/ccuracy.sh
```
Comment (pdxjohnny) [inline model/tensorflow_hub/dffml_model_tensorflow_hub/text_classifier.py:340]:
```suggestion
    .. literalinclude:: /../model/tensorflow_hub/examples/tfhub_text_classifier/accuracy.sh
```

File: model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        # Path to target file
+        filepath = os.path.join(os.path.dirname(__file__), filename)
+        # Capture output
+        stdout = subprocess.check_output([sys.executable, filepath])
+        lines = stdout.decode().split("\n")
+        # Check the Accuracy
+        self.assertEqual(lines[0], "Accuracy: 1.0")
+        # Check the TARGET
+        self.assertIsInstance(
+            round(ast.literal_eval(lines[1])["Sentiment"]), int
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:38]:
```suggestion
            round(ast.literal_eval(lines[1])["sentiment"]), int
```

File: model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+class TestExample(unittest.TestCase):
+    def python_test(self, filename):
+        # Path to target file
+        filepath = os.path.join(os.path.dirname(__file__), filename)
+        # Capture output
+        stdout = subprocess.check_output([sys.executable, filepath])
+        lines = stdout.decode().split("\n")
+        # Check the Accuracy
+        self.assertEqual(lines[0], "Accuracy: 1.0")
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:35]: I don't think this line will pass with the way the Python script is written at the moment

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    predict=DefFeature("Sentiment", int, 1),
+    classifications=[0, 1, 2],
+    clstype=int,
+)
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:10]:
```suggestion
)

```

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+# train the model
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:11]:
```suggestion
# Train the model
```

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+train(model, "train.csv")
+# Assess accuracy (alternate way of specifying data source)
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:13]:
```suggestion

# Assess accuracy (alternate way of specifying data source)
```

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+print(accuracy(model, CSVSource(filename="test.csv")))
+# Make prediction
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:15]:
```suggestion

# Make prediction
```

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+for i, features, prediction in predict(
+    model,
+    {"Sentence": "This track is horrible"},
+    {"Sentence": "Deforestation is bad"},
+    {"Sentence": "This restaurant is amazing"},
+):
+    features["Sentiment"] = prediction["Sentiment"]["value"]
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:22]:
```suggestion
    features["sentiment"] = prediction["sentiment"]["value"]
```

File: model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py (hunk 1 above)
Comment (pdxjohnny) [reply model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:12]: It's saying there are not records to train on because there are no records with matching features. Because the features are not the same capitalization

Comment (mHash1m) [reply model/tensorflow_hub/examples/tfhub_text_classifier/textclassifier.py:12]: Ah! I see. Thank you!

File: model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py (hunk 7 above)
Comment (mHash1m) [reply model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:35]: Yes, it doesn't

Comment (mHash1m) [reply model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:35]: The following output pops at running train which is bugging the tests. I'll look for a way around it in the tests.

`DEBUG:dffml.TextClassifierContext:Training on features: ['sentence']
INFO:dffml.TextClassifierContext:------ Record Data ------
INFO:dffml.TextClassifierContext:x_cols:    6
INFO:dffml.TextClassifierContext:y_cols:    6
INFO:dffml.TextClassifierContext:-----------------------
Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_text (InputLayer)      [(None,)]                 0         
_________________________________________________________________
keras_layer (KerasLayer)     (None, 20)                389380    
_________________________________________________________________
dense (Dense)                (None, 256)               5376      
_________________________________________________________________
dense_1 (Dense)              (None, 3)                 771       
=================================================================
Total params: 395,527
Trainable params: 395,527
Non-trainable params: 0
_________________________________________________________________
Train on 6 samples
Epoch 1/30
6/6 [==============================] - 0s 81ms/sample - loss: 3.7750e-07 - accuracy: 1.0000
`

Comment (pdxjohnny) [reply model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:35]: Hmmm, weird, ya probably best to just skip to parsing the last lines in stdout

File: model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+        # Path to target file
+        filepath = os.path.join(os.path.dirname(__file__), filename)
+
+        # Capture output
+        stdout = subprocess.check_output([sys.executable, filepath])
+        lines = stdout.decode().split("\n")
+
+        # Check the Accuracy
+        self.assertEqual(lines[-3], "Accuracy: 0.6666666865348816")
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:38]: Let's not do an exact check on accuracy. Instead let's split this line and check the the first part is "Accuracy:" and then ext part is a float.

Comment (mHash1m) [reply model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py]: Oh, alright.

File: model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+        # Path to target file
+        filepath = os.path.join(os.path.dirname(__file__), filename)
+
+        # Capture output
+        stdout = subprocess.check_output([sys.executable, filepath])
+        lines = stdout.decode().split("\n")
+
+        # Check the Accuracy
+        self.assertRegex(lines[-3], r"Accuracy:  [-+]?\d*\.?\d+|\d+")
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:38]: nice

File: model/tensorflow_hub/examples/tfhub_text_classifier/predict.sh [hunk 15]
```diff
@@ -0,0 +1,11 @@
+dffml predict all \
+  -model text_classifier \
+  -model-predict sentiment:int:1 \
+  -model-classifications 0 1 \
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/predict.sh:4]: The issue is because we have a different number of classifications, when we go to do the accuracy command it's loading a different model.

File: model/tensorflow_hub/examples/tfhub_text_classifier/train.sh [hunk 16]
```diff
@@ -0,0 +1,14 @@
+dffml train \
+  -model text_classifier \
+  -model-epochs 30 \
+  -model-predict sentiment:int:1 \
+  -model-classifications 0 1 2 \
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/train.sh:5]:
```suggestion
  -model-classifications 0 1 \
```

File: model/tensorflow_hub/examples/tfhub_text_classifier/predict.sh (hunk 15 above)
Comment (pdxjohnny) [reply model/tensorflow_hub/examples/tfhub_text_classifier/predict.sh:4]: https://github.com/intel/dffml/blob/1635ed60627139f3cb50d6a8e5fbb3fcd678d1e4/model/tensorflow_hub/dffml_model_tensorflow_hub/text_classifier.py#L119-L124

File: model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+            # Run training
+            subprocess.check_output(["bash", sh_filepath("train.sh")])
+
+            # Check the Accuracy
+            stdout = subprocess.check_output(
+                ["bash", sh_filepath("accuracy.sh")]
+            )
+            lines = stdout.decode().split("\n")
+            self.assertRegex(lines[-2], r"Accuracy:  [-+]?\d*\.?\d+|\d+")
```
Comment (pdxjohnny) [inline model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:60]: This check should probably just be a regex on the last line output, since that should be the float value when run via the command line

```suggestion
            self.assertRegex(lines[-1], r"[-+]?\d*\.?\d+|\d+")
```

Comment (pdxjohnny) [reply model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:60]: > I think it should be the last line but you may have been right with `-2`

File: model/tensorflow_hub/examples/tfhub_text_classifier/predict.sh (hunk 15 above)
Comment (mHash1m) [reply model/tensorflow_hub/examples/tfhub_text_classifier/predict.sh:4]: Ah, damn. I was playing with the model wondering why do we not have the 3 classifications there and I let that slip :facepalm:

File: model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py (hunk 17 above)
Comment (mHash1m) [reply model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:60]: yeah I am using the last line since there's an empty line stored at the end so it's -2

Comment (mHash1m) [reply model/tensorflow_hub/examples/tfhub_text_classifier/test_textclassifier.py:60]: or in this case we can make it simpler and just fo with 1 :smile:

**Comment by mHash1m** [comment]
@pdxjohnny  Sure seems like you did more work than I on this one :D Very helpful tho! Thanks for the review.

File: model/tensorflow_hub/examples/tfhub_text_classifier/predict.sh (hunk 15 above)
Comment (pdxjohnny) [reply model/tensorflow_hub/examples/tfhub_text_classifier/predict.sh:4]: This happens to me ALL THE TIME I sometimes wonder if we should ditch the hashing for a more explicit way. Let's bring it up in the sync meeting tomorrow

**Comment by pdxjohnny** [comment]
> @pdxjohnny Sure seems like you did more work than I on this one :D Very helpful tho! Thanks for the review.

lol, I don't know about that. This model has been confusing for a few of us now, Great to finally have an example. Thanks!
